scripts/main/lychee.js

```diff
@@ -875,38 +1120,68 @@ lychee.setMode = function (mode) {
 			.unbind(["command+backspace", "ctrl+backspace"])
 			.unbind(["command+a", "ctrl+a"]);
 	}
-	if (!lychee.rights.is_admin) {
-		$("#button_users, #button_logs, #button_diagnostics").remove();
+	if (!lychee.rights.is_admin || mode === "view" || mode === "frame") {
+		$("#button_users, #button_logs, #button_diagnostics").hide();
 	}
 
+	const bodyClasses = document.querySelector("body").classList;
+
 	if (mode === "logged_in") {
+		if (!bodyClasses.contains("mode-gallery")) {
+			bodyClasses.replace("mode-none", "mode-gallery");
+			bodyClasses.replace("mode-frame", "mode-gallery");
+			bodyClasses.replace("mode-view", "mode-gallery");
+		}
 		// After login the keyboard short-cuts to login by password (l) and
 		// by key (k) are not required anymore, so we unbind them.
 		Mousetrap.unbind(["l"]).unbind(["k"]);
 
 		// The code searches by class, so remove the other instance.
-		$(".header__search, .header__clear", ".header__toolbar--public").remove();
+		$(".header__search, .header__clear", ".header__toolbar--public").hide();
 
 		if (!lychee.editor_enabled) {
-			$("#button_rotate_cwise").remove();
-			$("#button_rotate_ccwise").remove();
+			$("#button_rotate_cwise").hide();
+			$("#button_rotate_ccwise").hide();
 		}
 		return;
 	} else {
-		$(".header__search, .header__clear", ".header__toolbar--albums").remove();
-		$("#button_rotate_cwise").remove();
-		$("#button_rotate_ccwise").remove();
+		$(".header__search, .header__clear", ".header__toolbar--albums").hide();
+		$("#button_rotate_cwise").hide();
+		$("#button_rotate_ccwise").hide();
 	}
 
-	$("#button_settings, .header__divider, .leftMenu").remove();
+	$("#button_settings, .header__divider, .leftMenu").hide();
 
 	if (mode === "public") {
+		if (!bodyClasses.contains("mode-gallery")) {
+			bodyClasses.replace("mode-none", "mode-gallery");
+			bodyClasses.replace("mode-frame", "mode-gallery");
+			bodyClasses.replace("mode-view", "mode-gallery");
+		}
 		lychee.publicMode = true;
 	} else if (mode === "view") {
+		if (!bodyClasses.contains("mode-view")) {
+			bodyClasses.replace("mode-none", "mode-view");
+			bodyClasses.replace("mode-frame", "mode-view");
+			bodyClasses.replace("mode-gallery", "mode-view");
+		}
 		Mousetrap.unbind(["esc", "command+up"]);
 
-		$("#button_back, a#next, a#previous").remove();
-		$(".no_content").remove();
+		$("#button_back, a#next, a#previous").hide();
+		$(".no_content").hide();
+
+		lychee.publicMode = true;
+		lychee.viewMode = true;
+	} else if (mode === "frame") {
+		if (!bodyClasses.contains("mode-frame")) {
+			bodyClasses.replace("mode-none", "mode-frame");
+			bodyClasses.replace("mode-view", "mode-frame");
+			bodyClasses.replace("mode-gallery", "mode-frame");
+		}
+		Mousetrap.unbind(["esc", "command+up"]);
+
+		$("#button_back, a#next, a#previous").hide();
+		$(".no_content").hide();
 
 		lychee.publicMode = true;
 		lychee.viewMode = true;
```

Comment:
Should `viewMode` really be set to `true` here?

Reply:
I am not sure. At least there does not seem to be a technical necessity for it. The variable `lychee.viewMode` is only queried in three places and none of them belongs to code which is executed during frame mode. Hence, the value does not matter.

However, I did it that way to keep the code consistent with the code in the `else if`-branch for the view mode earlier (13 lines above). See
```php
} else if (mode === "view") {
	if (!bodyClasses.contains("mode-view")) {
		bodyClasses.replace("mode-none", "mode-view");
		bodyClasses.replace("mode-frame", "mode-view");
		bodyClasses.replace("mode-gallery", "mode-view");
	}
	Mousetrap.unbind(["esc", "command+up"]);

	$("#button_back, a#next, a#previous").hide();
	$(".no_content").hide();

	lychee.publicMode = true;
	lychee.viewMode = true;
} else if (mode === "frame") {
	if (!bodyClasses.contains("mode-frame")) {
		bodyClasses.replace("mode-none", "mode-frame");
		bodyClasses.replace("mode-view", "mode-frame");
		bodyClasses.replace("mode-gallery", "mode-frame");
	}
	Mousetrap.unbind(["esc", "command+up"]);

	$("#button_back, a#next, a#previous").hide();
	$(".no_content").hide();

	lychee.publicMode = true;
	lychee.viewMode = true;
}
```

Except for the CSS classes on the `<body>` both branches are identical. In fact, it is also unnecessary to hide the buttons `a#previos` and `a#next` in frame mode, because these buttons will be completely covered by the background of the frame anyway. Hence, the last `else-if`-branch may even be simplified further. The only essential aspect is to unbind the "ESC" key.

However, in my perspective the view mode and the frame mode have a lot in common. They both show a large photo in the center and disallow most interaction with a user. In the long run both modes might share more code (for the image in the center) than they do at the moment. In order to stress that similarity, I considered it helpful to keep both branches as closely together as possible. That is why the code is as it is. But of course this is arguable.

Do you want me to remove that code?

Reply:
I have indeed wondered why we even try to hide some buttons that won't be visible anyway. We do need to make sure that the keyboard shortcuts don't work, but other than that...

But since the code is already there, I'm not overly bothered by it. It just seems... redundant.

Reply:
OK, so now that I tried this in person, I think we definitely need to do more with the view mode, because when I'm logged in, I'm seeing a whole lot of extra buttons:
![1](https://user-images.githubusercontent.com/16415200/197251092-3d92b2e5-9a1d-48b0-9008-b1c503b29c70.png)
Some of them work (rotation), some don't (starring). I think they should all be removed. Even when I'm not logged in, the little triangle to the right of the title is still there, and it shouldn't.

Reply:
That should definitely do not happen. And it did not happen when I finished my work on this PR. I fear this is another victim of the many merges and rebases to keep up with master. Sometimes I really hate Git as it is too brave sometimes. It merges things "successfully" but in fact some outdated, wrong, etc. pp. code slips in.

I need to check what went wrong here.

Reply:
I fixed the problem (hopefully) in https://github.com/LycheeOrg/Lychee-front/pull/323/commits/acd3b710e306f95aded8c643e2df7d6ed6669966.

I also remember the initial problem which has led to the previous unfortunate situation.

Initially, when the "view" mode had been implemented separately, the view mode provided a toolbar called `header__toolbar--public`. This toolbar contained the share and info button for a single photo. However, this name had already been taken in "gallery" mode for the toolbar for public albums. Moreover, the interdependent view mode defined the toolbar as
```html
<div class="header__toolbar header__toolbar--public header__toolbar--visible">
	<a class="header__title"></a>
	<a class="button button--share" id="button_share" title="Share Photo"><svg class="iconic ionicons"><use link:href="#share-ion"></use></svg></a>
	<a class="button button--info" id="button_info" title="About Photo"><svg class="iconic"><use xlink:href="#info"></use></svg></a>
</div>
```
but at the same time, the gallery mode contained a toolbar `header__toolbar--photo`
```html
<div class="header__toolbar header__toolbar--photo">
	<a class="button" id="button_back"><svg class="iconic"><use xlink:href="#chevron-left" /></svg></a>
	<a class="header__title"></a>
	<!-- ... -->
	<a class="button button--share" id="button_share"><svg class="iconic ionicons"><use xlink:href="#share-ion" /></svg></a>
	<a class="button button--info" id="button_info"><svg class="iconic"><use xlink:href="#info" /></svg></a>
	<!-- ... -->
</div>
```
which (among other) defined two buttons with the same IDs `button_share` and `button_info`. This was the point when I decided to drop the previously independent toolbar `header__toolbar--public` and only use `header__toolbar--photo`.

> Some of them work (rotation), some don't (starring). I think they should all be removed.

In the long run, I would object. IMHO, there should not be an independent view mode at all, but simply a single photo view whose toolbar adopts to the actions which are currently available and takes the authenticated user into account. I thought it would be nice to already push things into that direction, but it isn't that easy, because certain actions don't work in view mode and it is not possible to make them work without refactoring a lot more. This is something which can only be implemented after https://github.com/LycheeOrg/Lychee-front/issues/343.

For the time being, I agree that the non-functional buttons should simply be removed. This is the easiest solution for now. I "hacked" the necessary code into `header.setMode`. This way, "view mode" and "photo mode" still use the same toolbar, i.e. `header__toolbar--photo`, and we don't have the problem of duplicated IDs, but we hide buttons if `lychee.publicMode` is set. This is not a great solution, but requires the least lines of additional code. Moreover, we already do similar things for the album and map toolbar, hence it is even somewhat consistent.

Reply:
Agreed. We should make the photo mode work in both cases.